jmx-metrics/src/main/groovy/io/opentelemetry/contrib/jmxmetrics/InstrumentHelper.groovy

```diff
@@ -58,81 +62,54 @@ class InstrumentHelper {
      * @param instrument - The {@link io.opentelemetry.api.metrics.Instrument}-producing {@link OtelHelper} method pointer:
      *        (e.g. new OtelHelper().&doubleValueRecorder)
      */
-    InstrumentHelper(MBeanHelper mBeanHelper, String instrumentName, String description, String unit, Map<String, Closure<?>> labelFuncs, Map<String, Map<String, Closure<?>>> MBeanAttributes, Closure<?> instrument) {
+    InstrumentHelper(MBeanHelper mBeanHelper, String instrumentName, String description, String unit, Map<String, Closure<?>> labelFuncs, Map<String, Map<String, Closure<?>>> MBeanAttributes, Closure<?> instrument, GroovyMetricEnvironment metricEnvironment) {
         this.mBeanHelper = mBeanHelper
         this.instrumentName = instrumentName
         this.description = description
         this.unit = unit
         this.labelFuncs = labelFuncs
         this.mBeanAttributes = MBeanAttributes
         this.instrument = instrument
+        this.metricEnvironment = metricEnvironment
     }
 
     void update() {
-        // Tuples of the form (mbean, attribute, value)
-        def values = mBeanHelper.getAttributes(mBeanAttributes.keySet())
-
-        // If there are no tuples with non-null value, return early
-        if (values.find {it.getV3() != null } == null) {
-            logger.warning("No valid value(s) for ${instrumentName} - ${mBeanHelper}.${mBeanAttributes.keySet().join(",")}")
-            return
+        def mbeans = mBeanHelper.getMBeans()
+        def compositeAttributes = []
+        def simpleAttributes = []
+        if (mbeans.size() == 0) {
+          return
         }
-
-        // Observer instruments need to have a single updater set at build time, so pool all
-        // update operations in a list of closures per instrument to be executed after all values
-        // are established, potentially as a single updater.  This is done because a single MBeanHelper
-        // can represent multiple MBeans (each with different values for an attribute) and the labelFuncs
-        // will create multiple datapoints from the same instrument identifiers.
-        def tupleToUpdates = [:] // tuple is of form (instrument, instrumentName, description, unit)
-
-        values.each { collectedValue ->
-            def mbean = collectedValue.getV1()
-            def attribute = collectedValue.getV2()
-            def value = collectedValue.getV3()
+        // Look at the first mbean collected to evaluate if the attributes requested are
+        // composite data types or simple. Composite types require different parsing to
+        // end up with multiple recorders in the same callback.
+        def bean = mbeans.first()
```

Comment:
Are we sure it's guaranteed the first will be reflective of the rest?

Reply:
it's not guaranteed 100%, but it would break metric conventions if some mbeans in a group were composite data and some were not (IE you would have `jvm.memory.pool` metrics & the `jvm.memory.pool.(committed|used|etc)` metrics if that instrument contained some composite & some simple.

We could add some complexity in order to split mbeans by type, or maybe we just log an error and bail if not all of the mbeans in an instrument match?

Reply:
Okay. As the batch callback needs to register all of its expected instruments, this might be a little complex, will take a pass at it.

Reply:
I believe this is improved, let me know!